cpp/src/arrow/type.h

```diff
@@ -677,7 +678,8 @@ class ARROW_EXPORT DictionaryType : public FixedWidthType {
 
 class ARROW_EXPORT Schema {
  public:
-  explicit Schema(const std::vector<std::shared_ptr<Field>>& fields);
+  explicit Schema(const std::vector<std::shared_ptr<Field>>& fields,
+      const KeyValueMetadata& custom_metadata = KeyValueMetadata());
```

Comment:
this suggests we might want to make the API

`Schema(..., std::unique_ptr<KeyValueMetadata> custom_metadata)` and
`Schema(..., std::unique_ptr<const KeyValueMetadata> custom_metadata)` and

with the storage in this class as `std::unique_ptr<const KeyValueMetadata>`. I think you can move `std::unique_ptr<T>` into `std::unique_ptr<const T>`

We could further make KeyValueMetadata non-copyable. Still `custom_metadata()` returns `const KeyValueMetadata&`. We could do this in a follow up patch